### PR TITLE
i#5190: drcachesim child stats include all lower levels

### LINKED ISSUE #5190: drcachesim child cache stats should include all lower levels
The child (==lower level) cache stats in drcachesim today are only for the immediately lower level.  This is a holdover from the original implementation which didn't support other than 2 levels.  For multiple levels, one would expect the child stats to include the sum of all the lower levels.

### PR DESCRIPTION
**Pull request by derekbruening**
The child (==lower level) cache stats in drcachesim today are only for
the immediately lower level. This is a holdover from the original
implementation which didn't support other than 2 levels. For multiple
levels, one would expect the child stats to include the sum of all the
lower levels.  We fix that here and add a unit test.

The difference is clear in the config file test:
  LLC stats:
     Child hits:                        147
=>
  LLC stats:
     Child hits:                    174,695

But to avoid flakiness we do not update its regex and rely on the unit
test.

Fixes #5190

### CODE REVIEW

File: clients/drcachesim/tests/drcachesim_unit_tests.cpp [hunk 1]
```diff
@@ -235,6 +235,64 @@ unit_test_compulsory_misses()
...
+
+    const int num_accesses = 16;
+    for (int i = 0; i < num_accesses; i++) {
+        ref.data.addr = 64 + i;
+        if (!cache_sim.process_memref(ref)) {
+            std::cerr << "drcachesim unit_test_child_hits failed: "
+                      << cache_sim.get_error_string() << "\n";
+            exit(1);
+        }
+    }
```
Comment (abhinav92003) [inline clients/drcachesim/tests/drcachesim_unit_tests.cpp:291]: The cache line size is presumably 64. So `assert(num_accesses < 64)` too?

File: clients/drcachesim/tests/drcachesim_unit_tests.cpp [hunk 2]
```diff
@@ -235,6 +235,64 @@ unit_test_compulsory_misses()
...
+    assert(cache_sim.get_cache_metric(metric_name_t::CHILD_HITS, 3, 0) ==
```
Comment (abhinav92003) [inline clients/drcachesim/tests/drcachesim_unit_tests.cpp:292]: For a slightly stronger check, maybe add  `&& cache_sim.get_cache_metric(metric_name_t::CHILD_HITS, 1, 0)`

File: clients/drcachesim/simulator/caching_device.cpp [hunk 3]
```diff
@@ -185,7 +185,8 @@ caching_device_t::request(const memref_t &memref_in)
             missed = true;
             // If no parent we assume we get the data from main memory
             if (parent_ != NULL) {
-                parent_->stats_->child_access(memref, false, cache_block);
+                for (caching_device_t *up = parent_; up != nullptr; up = up->parent_)
```
Comment (abhinav92003) [inline clients/drcachesim/simulator/caching_device.cpp:188]: (Optional) Worth extracting this loop into a separate routine? `update_child_access`?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tests/drcachesim_unit_tests.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tests/drcachesim_unit_tests.cpp:291]: See the `line_size` right above.  Added a comment.  An assert seems redundant.  If something is messed up the stats will be off in any case.

File: clients/drcachesim/tests/drcachesim_unit_tests.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tests/drcachesim_unit_tests.cpp:292]: Added several more stats asserts.

File: clients/drcachesim/simulator/caching_device.cpp (hunk 3 above)
Comment (derekbruening) [reply clients/drcachesim/simulator/caching_device.cpp:188]: Done as part of changing the logic to propagate hits all the way but misses just one level.

Comment (derekbruening) [reply clients/drcachesim/simulator/caching_device.cpp:188]: Unfortunately there's no way to test the miss logic b/c dr$sim does not care about child misses: it's something we want for external uses that subclasses this.  Prob sufficient to have just external tests.

**Pull request merged**